### ISSUE TITLE
attention mask during training

### ISSUE THREAD

**Issue by kzkedzierska** [bowang-lab/scGPT#47]
Hi, 

I've been thinking some more about the attention masking and reading up on the flash attention itself. 

If I understand correctly, a custom mask needs to be supplied to the attention to replicate Fig S1A:

![Supplementary Figure 1 from scGPT v2 preprint](https://www.biorxiv.org/content/biorxiv/early/2023/07/02/2023.04.30.538439/F7.large.jpg?width=800&height=600&carousel=1)

If I read this correctly, this is to make all query independent of the unknown genes, except those individual unknown genes, right?

However, my understanding is that [flash-attention](https://github.com/Dao-AILab/flash-attention/tree/v1.0.4) (`FlashMHA` which calls `FlashAttention`) only supports `key_padding_mask`. Do you use a different attention during training? Or do you use `key_padding_mask` in some more complex way?

I will be very grateful for your response, I'm keen on understanding the model better!

**Comment by subercui** [comment]
Hi @kzkedzierska , sorry for the delay! We did use customized implementation during pre-training. The release of data collection and pretraining code took longer than we expected, and I was so occupied last week. We will try our best to release them by tomorrow evening. Then it will be clearer to answer the question, and I will explain with the code.

**Comment by kzkedzierska** [comment]
@subercui is there any ETA on this? I'm trying to understand the model but without the training implementation I am finding it challenging.

**Comment by subercui** [comment]
Hi @kzkedzierska , thank you for your patience sincerely. The current challenge is merely the workload of merging some legacy code with different naming and other misc, while we are also working toward other requests. How about I can make up a separate temporary branch tonight with relevant code? I will try to merge the code later after providing an explanation to the questions you have.

**Comment by kzkedzierska** [comment]
Thanks for replying! That sounds great, thank you!

**Comment by subercui** [comment]
Hi, I have uploaded the code in the dev-temp branch. Regarding your initial questions:

1. Yes, the design would make each query independent of all other unknown genes
2. We did make our own implementation to make the customized attention work. Currently, it is a combination of pytorch MHA and flash-attn MHA. Please see the implementation here https://github.com/bowang-lab/scGPT/blob/dev-temp/scgpt/model/flash_layers.py#L19-L189 . To summarize, the upper part of the attention map would have a regular square shape, so we used flash-attn for it. We used the Pytorch MHA with custom attention mask input for the other part.

**Comment by kzkedzierska** [comment]
Thank you posting the code!

I will be very grateful if you could please help me understand the code. 

Would you mind expanding on what do you mean by `pcpt_gene` and `gen_gene`? How do they relate to mask visualization attached earlier?

Could you comment on how the pre-trained model relates to the one that was used so far in the forward passes in the notebooks? Currently the code contains 4 different variants of the Transformer Layers per code [here](https://github.com/kzkedzierska/scGPT/blob/4068d67caaac1e28d56964da68e0214817e38428/scgpt/model/model.py#L109-L139):

1. if `use_generative_training` is set to `True` it's the custom `FlashscGPTLayer` and `FlashscGTPGenerator`;
2. if  using`use_ fast_transformer` is set to `True` and  `fast_transformer_backend` set to `linear` it's the TransformerLayers created by `FastTransformerEncoderWrapper`.
3. if using `use_ fast_transformer` is set to `True` and  `fast_transformer_backend` set to `flash` (as shown in the `args.json` file for the pretrained models) it's `FlashTransformerEncoderLayer` and `TransformerEncoder`; and
4. if neither `use_generative_training` nor `use_ fast_transformer` are set to True then it's the vanilla Pytorch Transformer. 

Is there an equivalency between those architectures? As in, if you pretrain with the 1. architecture can I use the model in inference mode with architecture 3. or 4. (should I, for example, not have access to the GPU supported by flash attention supported)? 

And am I correct to think that the custom implementation of 1. would be an equivalent to 4. with providing a custom mask as per the illustration earlier in the `attn_mask` argument of the PyTorch `MultiheadAttention`? Or is there some other additional step I am missing?

**Comment by zhaolin64** [comment]
> Thank you posting the code!
> 
> I will be very grateful if you could please help me understand the code.
> 
> Would you mind expanding on what do you mean by `pcpt_gene` and `gen_gene`? How do they relate to mask visualization attached earlier?
> 
> Could you comment on how the pre-trained model relates to the one that was used so far in the forward passes in the notebooks? Currently the code contains 4 different variants of the Transformer Layers per code [here](https://github.com/kzkedzierska/scGPT/blob/4068d67caaac1e28d56964da68e0214817e38428/scgpt/model/model.py#L109-L139):
> 
> 1. if `use_generative_training` is set to `True` it's the custom `FlashscGPTLayer` and `FlashscGTPGenerator`;
> 2. if  using `use_fast_transformer` is set to `True` and  `fast_transformer_backend` set to `linear` it's the TransformerLayers created by `FastTransformerEncoderWrapper`.
> 3. if using `use_fast_transformer` is set to `True` and  `fast_transformer_backend` set to `flash` (as shown in the `args.json` file for the pretrained models) it's `FlashTransformerEncoderLayer` and `TransformerEncoder`; and
> 4. if neither `use_generative_training` nor `use_fast_transformer` are set to True then it's the vanilla Pytorch Transformer.
> 
> Is there an equivalency between those architectures? As in, if you pretrain with the 1. architecture can I use the model in inference mode with architecture 3. or 4. (should I, for example, not have access to the GPU supported by flash attention supported)?
> 
> And am I correct to think that the custom implementation of 1. would be an equivalent to 4. with providing a custom mask as per the illustration earlier in the `attn_mask` argument of the PyTorch `MultiheadAttention`? Or is there some other additional step I am missing?

Hello, I also encountered the same issue when trying to replicate this work, especially regarding pcpt_gene. Do you understand it now? Looking forward to your reply, thank you.
